### PR TITLE
Components

### PR DESCRIPTION
**Pull request by mhajij**
1- "components" is now a module that supports CHG, CC and CCC. No need to have these functions for each class as the code repeats. 
2- "distance" is now a module that supports CHG, CC and CCC. No need to have these functions for each class as the code repeats. 
3- distance measures are added also as new module (cell diameter, node diameter, diameter, etc).  No need to have these functions for each class as the code repeats.

The above is more consistent with NetworkX in terms of API as well as folder structure.

### CODE REVIEW

File: test/algorithms/test_components.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+        # Test the function with cells=False and return_singletons=True
+        result = list(connected_components(CX, cells=True, return_singletons=True))
+        expected_result = [
+            {(2, 3), (2, 3, 4), (2, 4), (3, 4)},
+            {(5, 6), (5, 6, 7), (5, 7), (6, 7)},
+        ]
+        assert result == expected_result
+
+        CX = CellComplex()
```
Comment (ninamiolane) [inline test/algorithms/test_components.py:96]: Our API uses CC as the abbreviation for cell complex. -> Update?

Comment (ninamiolane) [reply test/algorithms/test_components.py:96]: Here and everywhere

File: toponetx/algorithms/components.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+__all__ = [
+    "s_connected_components",
+    "s_component_subcomplexes",
+    "connected_components",
+    "connected_component_subcomplexes",
+]
+
+
+def s_connected_components(
+    cplex, s: int = 1, cells: bool = True, return_singletons: bool = False
```
Comment (ninamiolane) [inline toponetx/algorithms/components.py:24]: Our API does not use `cplex` but `complex`. Update? Here and everywhere.

File: toponetx/algorithms/components.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    >>> CC.add_cell([2,3,4],rank=2)
+    >>> CC.add_cell([5,6,7],rank=2)
+    >>> list(connected_components(CC,cells=False))
+    >>> CC.add_cell([4,5],rank=1)
+    >>> list(CC.connected_components(CC,cells=False))
+    """
+    return s_connected_components(cmplex, s=1, cells=cells, return_singletons=True)
+
+
+def connected_component_subcomplexes(cplex, return_singletons=True):
```
Comment (ninamiolane) [inline toponetx/algorithms/components.py:179]: Incomplete docstrings: Need Parameters section

File: toponetx/algorithms/components.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            cplex, s=s, cells=cells, return_singletons=return_singletons
+        )
+    ):
+        if cells:
+            yield cplex.restrict_to_cells(list(c))
+        else:
+            yield cplex.restrict_to_nodes(list(c))
+
+
+def connected_components(cmplex, cells=False, return_singletons=True):
```
Comment (ninamiolane) [inline toponetx/algorithms/components.py:154]: Incomplete docstrings: Need Parameters section

File: toponetx/algorithms/components.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def s_component_subcomplexes(cplex, s=1, cells=True, return_singletons=False):
+    """Return a generator for the induced subcomplexes of s_connected components.
+
+    Removes singletons unless return_singletons is set to True.
+
+    Parameters
+    ----------
+    s : int, optional
```
Comment (ninamiolane) [inline toponetx/algorithms/components.py:117]: Need parameter description

File: toponetx/algorithms/components.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def s_connected_components(
+    cplex, s: int = 1, cells: bool = True, return_singletons: bool = False
+):
+    """Return generator for the s-connected components.
+
+    Parameters
+    ----------
+    s : int, optional
```
Comment (ninamiolane) [inline toponetx/algorithms/components.py:30]: Need parameter description

File: toponetx/algorithms/distance.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    G = nx.from_scipy_sparse_array(A)
+    try:
+        path = nx.shortest_path_length(G, rowdict[source], rowdict[target])
+        return path
+    except Exception:
+        warn(f"No {s}-path between {source} and {target}")
+        return np.inf
+
+
+def cell_distance(cmplex, source, target, s=1):
```
Comment (ninamiolane) [inline toponetx/algorithms/distance.py:80]: Our API does not use `cmplex` but `complex`: update?

File: toponetx/algorithms/distance.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    """Return the shortest s-walk distance between two cells in the cell complex.
+
+    Parameters
```
Comment (ninamiolane) [inline toponetx/algorithms/distance.py:83]: Parameters section is not complete: missing complex

File: toponetx/algorithms/distance.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    ----------
+    source : cell.uid or cell
+        a cell in the cell complex
```
Comment (ninamiolane) [inline toponetx/algorithms/distance.py:86]: This does not respect our docstrings conventions, and - as a result - will not look clean on the doc website: update?
Cf: https://pyt-team.github.io/topomodelx/contributing/index.html#write-documentation

File: toponetx/algorithms/distance.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    if isinstance(source, (Cell, HyperEdge)):
+        source = source.elements
+    if isinstance(target, (Cell, HyperEdge)):
+        target = target.elements
+    if isinstance(cmplex, CellComplex):
+        if isinstance(source, Iterable):
+            source = tuple(source)
+        if isinstance(target, Iterable):
+            target = tuple(target)
+    cell_dict, A = cmplex.all_cell_to_node_codjacnecy_matrix(s=s, index=True)
```
Comment (ninamiolane) [inline toponetx/algorithms/distance.py:137]: typo: codjacnecy

File: toponetx/algorithms/distance_measures.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        diamc = nx.diameter(G.subgraph(c))
+        temp = set()
+        for e in c:
+            temp.add(node_dict[e])
+        comps.append(temp)
+        diams.append(diamc)
+    return diams, comps
+
+
+def cell_diameters(cplex, s=1):
```
Comment (ninamiolane) [inline toponetx/algorithms/distance_measures.py:59]: Ditto

File: toponetx/algorithms/distance_measures.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    """Return the cell diameters of the s_cell_connected component subgraphs.
+
+    Parameters
```
Comment (ninamiolane) [inline toponetx/algorithms/distance_measures.py:62]: Erroneous Parameters section

**Review by ninamiolane — CHANGES_REQUESTED** [review]
See several API and documentation errors.

File: test/algorithms/test_components.py (hunk 1 above)
Comment (mhajij) [reply test/algorithms/test_components.py:96]: done

File: toponetx/algorithms/components.py (hunk 2 above)
Comment (mhajij) [reply toponetx/algorithms/components.py]: done

File: toponetx/algorithms/components.py (hunk 3 above)
Comment (mhajij) [reply toponetx/algorithms/components.py]: done

File: toponetx/algorithms/components.py (hunk 4 above)
Comment (mhajij) [reply toponetx/algorithms/components.py]: done

File: toponetx/algorithms/components.py (hunk 5 above)
Comment (mhajij) [reply toponetx/algorithms/components.py:119]: done

**Review by ninamiolane — CHANGES_REQUESTED** [review]
Several comments have not been addressed.
This includes: Typo: codjacnecy and docstrings comments.

**Comment by ninamiolane** [comment]
Tests are failing. Please fix the tests before marking a PR ready for review.

**Comment by mhajij** [comment]
> Tests are failing. Please fix the tests before marking a PR ready for review.

now done. All tests pass.

File: toponetx/algorithms/distance.py (hunk 9 above)
Comment (mhajij) [reply toponetx/algorithms/distance.py]: done

File: toponetx/algorithms/distance_measures.py (hunk 11 above)
Comment (mhajij) [reply toponetx/algorithms/distance_measures.py]: done

File: toponetx/algorithms/distance_measures.py (hunk 12 above)
Comment (mhajij) [reply toponetx/algorithms/distance_measures.py:62]: done

File: toponetx/algorithms/distance.py (hunk 10 above)
Comment (mhajij) [reply toponetx/algorithms/distance.py]: done

File: toponetx/algorithms/distance.py (hunk 8 above)
Comment (mhajij) [reply toponetx/algorithms/distance.py:84]: done

File: toponetx/algorithms/components.py (hunk 4 above)
Comment (mhajij) [reply toponetx/algorithms/components.py]: done

File: toponetx/algorithms/components.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+    adjacent nodes in the sequence share s cells. If s=1 these are the
+    path components of the cell complex .
+
+    Yields
+    ------
+    s_connected_components : iterator
+        Iterator returns sets of uids of the cells (or nodes) in the s-cells(node)
+        components of the cell complex.
+
+    Example
```
Comment (ffl096) [inline toponetx/algorithms/components.py:60]: NumPy's documentation style dictates that this headline should read `Examples` even if there is only one (same throughout this file).

File: toponetx/algorithms/components.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    >>> CC.add_cell([2,3,4],rank=2)
+    >>> CC.add_cell([5,6,7],rank=2)
+    >>> list(connected_components(CC,cells=False))
+    >>> CC.add_cell([4,5],rank=1)
+    >>> list(CC.connected_components(CC,cells=False))
+    """
+    return s_connected_components(complex, s=1, cells=cells, return_singletons=True)
+
+
+def connected_component_subcomplexes(complex, return_singletons=True):
```
Comment (ffl096) [inline toponetx/algorithms/components.py:195]: - Type hint for `complex` parameter is missing
- Add a return type

File: toponetx/algorithms/distance.py [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+from toponetx.classes.colored_hypergraph import ColoredHyperGraph
+from toponetx.classes.combinatorial_complex import CombinatorialComplex
+from toponetx.classes.hyperedge import HyperEdge
+from toponetx.classes.simplicial_complex import SimplicialComplex
+from toponetx.exception import TopoNetXError
+
+__all__ = ["distance", "cell_distance"]
+
+
+def distance(complex, source, target, s=1):
```
Comment (ffl096) [inline toponetx/algorithms/distance.py:19]: Missing type annotations

File: toponetx/algorithms/distance.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    source : a hashable represeting a node in the complex
+        A node in the input complex.
+    target : a hashable represeting a node
+        A node in the input complex.
+    s : int
+        the number of intersections between pairwise consecutive cells
+
+    Returns
+    -------
+    s_walk_distance : int
```
Comment (ffl096) [inline toponetx/algorithms/distance.py:35]: single return values do not have a name. Remove the s_connected_components : part (same throughout this file).

File: toponetx/algorithms/distance.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+    """Return shortest s-walk distance between two nodes in the cell complex.
+
+    Parameters
+    ----------
+    complex : Complex, supported complexes are cell/combintorial and hypegraphs
+    source : a hashable represeting a node in the complex
+        A node in the input complex.
+    target : a hashable represeting a node
+        A node in the input complex.
+    s : int
```
Comment (ffl096) [inline toponetx/algorithms/distance.py]:
```suggestion
    s : int, default=1
```

File: toponetx/algorithms/distance.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+    -----
+    The s-distance is the shortest s-walk length between the nodes.
+    An s-walk between nodes is a sequence of nodes that pairwise share
+    at least s cells. The length of the shortest s-walk is 1 less than
+    the number of nodes in the path sequence.
+
+    Uses the networkx shortest_path_length method on the graph
+    generated by the s-adjacency matrix.
+
+    Example
```
Comment (ffl096) [inline toponetx/algorithms/distance.py:51]: NumPy's documentation style dictates that this headline should read Examples even if there is only one (same throughout this file).

File: toponetx/algorithms/distance.py [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+    G = nx.from_scipy_sparse_array(A)
+    try:
+        path = nx.shortest_path_length(G, rowdict[source], rowdict[target])
+        return path
+    except Exception:
+        warn(f"No {s}-path between {source} and {target}")
+        return np.inf
+
+
+def cell_distance(complex, source, target, s=1):
```
Comment (ffl096) [inline toponetx/algorithms/distance.py:82]: Missing type annotations

File: toponetx/algorithms/distance_measures.py [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+from toponetx.classes.cell_complex import CellComplex
+from toponetx.classes.colored_hypergraph import ColoredHyperGraph
+from toponetx.classes.combinatorial_complex import CombinatorialComplex
+from toponetx.classes.simplicial_complex import SimplicialComplex
+from toponetx.exception import TopoNetXError
+
+__all__ = ["node_diameters", "cell_diameters", "diameter", "cell_diameter"]
+
+
+def node_diameters(complex):
```
Comment (ffl096) [inline toponetx/algorithms/distance_measures.py:17]: Missing type annotations

File: toponetx/algorithms/distance_measures.py [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+    """Return the node diameters of the connected components in cell complex.
+
+    Returns
+    -------
+    diameters : list
+        List of the diameters of the s-components.
+    components : list
+        List of the s-component nodes.
+
+    Example
```
Comment (ffl096) [inline toponetx/algorithms/distance_measures.py:27]: NumPy's documentation style dictates that this headline should read Examples even if there is only one (same throughout this file).

File: toponetx/algorithms/distance_measures.py [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+        diamc = nx.diameter(G.subgraph(c))
+        temp = set()
+        for e in c:
+            temp.add(node_dict[e])
+        comps.append(temp)
+        diams.append(diamc)
+    return diams, comps
+
+
+def cell_diameters(complex, s=1):
```
Comment (ffl096) [inline toponetx/algorithms/distance_measures.py:59]: Missing type annotations

File: toponetx/algorithms/distance_measures.py [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+        diamc = nx.diameter(G.subgraph(c))
+        temp = set()
+        for e in c:
+            temp.add(coldict[e])
+        comps.append(temp)
+        diams.append(diamc)
+    return diams, comps
+
+
+def diameter(complex) -> int:
```
Comment (ffl096) [inline toponetx/algorithms/distance_measures.py:106]: NumPy's documentation style dictates that this headline should read Examples even if there is only one (same throughout this file).

File: toponetx/algorithms/distance_measures.py [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+    if not isinstance(complex, (CellComplex, CombinatorialComplex, ColoredHyperGraph)):
+        raise ValueError("Input complex {complex} is not supported.")
+    A = complex.node_to_all_cell_adjacnecy_matrix()
+    G = nx.from_scipy_sparse_array(A)
+    if nx.is_connected(G):
+        return nx.diameter(G)
+    raise TopoNetXError("cc is not connected.")
+
+
+def cell_diameter(complex, s: int = None) -> int:
```
Comment (ffl096) [inline toponetx/algorithms/distance_measures.py:154]: Missing type annotation for `complex` parameter

File: toponetx/classes/cell_complex.py [hunk 25]
```diff
@@ -1377,18 +1377,24 @@ def node_to_all_cell_adjacnecy_matrix(
         scipy.sparse.csr.csc_matrix | tuple[dict, dict, scipy.sparse.csc_matrix]
             The adjacency matrix, if `index` is False, otherwise
             index of nodes, adjacency matrix, if 'index' is True
+
+        Note
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1381]:
```suggestion
        Notes
```

File: toponetx/classes/cell_complex.py [hunk 26]
```diff
@@ -1429,6 +1433,18 @@ def all_cell_to_node_codjacnecy_matrix(
         scipy.sparse.csr.csc_matrix | tuple[dict, dict, scipy.sparse.csc_matrix]
             The adjacency matrix, if `index` is False, otherwise
             index of cells, adjacency matrix, if 'index' is True
+
+        Note
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1437]:
```suggestion
        Notes
```

File: toponetx/classes/cell_complex.py [hunk 27]
```diff
@@ -1429,6 +1433,18 @@ def all_cell_to_node_codjacnecy_matrix(
...
+        ----
+        Two cells (1 dimensional or 2 dimensional) are s-coadjacent iff
+        they share a vertex.
+
+        Example
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1442]: NumPy's documentation style dictates that this headline should read Examples even if there is only one (same throughout this file).

File: toponetx/classes/cell_complex.py [hunk 28]
```diff
@@ -2048,6 +2064,37 @@ def to_combinatorial_complex(self):
...
+    def to_colored_hypergraph(self):
+        """Convert to colored hypergraph.
+
+        A cell complex is a type of combinatorial complex.
+        The rank of an element in a cell complex is its dimension, so vertices have rank 0,
+        edges have rank 1, and faces have rank 2.
+
+        Returns
+        -------
+        combinatorial_complex : CombinatorialComplex
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2076]:
```suggestion
        CombinatorialComplex
```

File: toponetx/classes/colored_hypergraph.py [hunk 29]
```diff
@@ -850,7 +851,7 @@ def remove_cells(self, cell_set) -> None:
 
         Returns
         -------
-        Colored Hypergraph : NestedColoredHyperGraph
+        Colored Hypergraph : ColoredHyperGraph
```
Comment (ffl096) [inline toponetx/classes/colored_hypergraph.py:854]:
```suggestion
        ColoredHyperGraph
```

File: toponetx/classes/colored_hypergraph.py [hunk 30]
```diff
@@ -1176,9 +1207,25 @@ def restrict_to_cells(self, cell_set, name: Optional[str] = None):
 
         Returns
         -------
-        new Colored Hypergraph : NestedColoredHyperGraph
+        new Colored Hypergraph : ColoredHyperGraph
```
Comment (ffl096) [inline toponetx/classes/colored_hypergraph.py:1210]:
```suggestion
        ColoredHyperGraph
```

File: toponetx/classes/colored_hypergraph.py [hunk 31]
```diff
@@ -1196,10 +1243,28 @@ def restrict_to_nodes(self, node_set, name: Optional[str] = None):
...
```
Comment (ffl096) [inline toponetx/classes/colored_hypergraph.py:1246]:
```suggestion
        ColoredHyperGraph
```

**Review by ffl096 — CHANGES_REQUESTED** [review]

File: toponetx/algorithms/components.py [hunk 32]
```diff
@@ -0,0 +1,220 @@
...
+            if not return_singletons:
+                if len(c) == 1:
+                    continue
+            if isinstance(complex, CellComplex):
+                yield {node_dict[n] for n in c}
+            else:
+                yield {tuple(node_dict[n])[0] for n in c}
+
+
+def s_component_subcomplexes(complex, s=1, cells=True, return_singletons=False):
```
[the remark this answers is not in the document]
Comment (mhajij) [reply toponetx/algorithms/components.py:113]: I have Yields instead of return iterator as the type. Isnt that the correct documentation ?

**Comment by ffl096** [comment]
The `complex` function parameters in this pull request shadows the built-in [`complex()`](https://docs.python.org/3/library/functions.html?highlight=complex#complex) function. Not too big of an issue, but something to be aware of. Maybe we can come up with a good replacement?

File: toponetx/algorithms/components.py [hunk 33]
```diff
@@ -0,0 +1,220 @@
...
+        subcomplexes equal to the cell complex restricted to each set of nodes(cells) in the
+        s-connected components or s-cell-connected components.
+    return_singletons : bool, optional
+                        When True, returns singletons connected components
+
+    Yields
+    ------
+    s_component_subcomplexes : iterator
+        Iterator returns subcomplexes generated by the cells (or nodes) in the
+        s-cell(node) components of cell complex.
```
[the remark this answers is not in the document]
Comment (mhajij) [reply toponetx/algorithms/components.py:135]: subcomplex is used mathematically here. As the restriction of the complex on its connected components. I made a change, let me know if you like it.

**Comment by mhajij** [comment]
> The `complex` function parameters in this pull request shadows the built-in [`complex()`](https://docs.python.org/3/library/functions.html?highlight=complex#complex) function. Not too big of an issue, but something to be aware of. Maybe we can come up with a good replacement?

I agree, @ninamiolane , your thoughts?

File: toponetx/classes/cell_complex.py (hunk 27 above)
Comment (mhajij) [reply toponetx/classes/cell_complex.py:1442]: I am not sure I understand this comment.

File: toponetx/algorithms/components.py (hunk 33 above)
[the remark this answers is not in the document]
Comment (ffl096) [reply toponetx/algorithms/components.py]: I think you misunderstood. That was not the issue here.

Look at the following example. The `Yields` headline already implies that you can iterate over the returned value of the function. You currently have one iterable "too many", i.e., currently your documentation implies that the user gets an iterable of iterables, which is not the case.
```python
def example():
    """
    Yields
    ------
    int
        This is the correct documentation for this function.
    """
    for i in range(5):
        yield i

File: toponetx/algorithms/components.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (ffl096) [reply toponetx/algorithms/components.py:114]: I didn't mean in the docstring but in the code. The parts after the `:` behind each parameter and behind `->` after the function signature :)

File: toponetx/algorithms/components.py (hunk 33 above)
[the remark this answers is not in the document]
Comment (mhajij) [reply toponetx/algorithms/components.py]: great, I will fix that.

[the remark this answers is not in the document]
Comment (mhajij) [reply toponetx/algorithms/components.py]: I think I fixed what you mean, Please let me know if it seems correct to you now.

File: toponetx/algorithms/components.py (hunk 13 above)
Comment (ninamiolane) [reply toponetx/algorithms/components.py:63]:
```
Examples
----------
```

File: toponetx/algorithms/distance.py (hunk 18 above)
Comment (ninamiolane) [reply toponetx/algorithms/distance.py:53]: +1

File: toponetx/algorithms/distance_measures.py [hunk 34]
```diff
@@ -0,0 +1,198 @@
...
+    complex : Complex
+        Supported complexes are cell/combintorial and hypegraphs.
+    s : int, optional
+        The number of intersections between pairwise consecutive cells.
+
+    Returns
+    -------
+    list of diameters : list
+        List of cell_diameters for s-cell component subcomplexes in the cell complex.
+
```
Comment (ninamiolane) [inline toponetx/algorithms/distance_measures.py:75]: No empty line

**Review by ninamiolane — APPROVED** [review]
LGTM! 

Minor comments:

fix typos: combintorial and hypegraphs (everywhere)
address @ffl096 comment on the Examples section

I agree that complex overshadows complex(). I think this is fine for now, since we do not compute with complex numbers.

**Comment by mhajij** [comment]
> LGTM

@ninamiolane 
I will fix typos in a different pull request to avoid more conflict.

**Comment by mhajij** [comment]
@ninamiolane this branch is diverging with other branches, please merge if you are happy with the current state.

**Comment by mhajij** [comment]
Closes to avoid further divergence

**Pull request merged**